core/src/main/java/com/sequenceiq/cloudbreak/service/stack/flow/StackSyncService.java

```diff
@@ -0,0 +1,141 @@
+import com.sequenceiq.cloudbreak.domain.CloudPlatform;
+import com.sequenceiq.cloudbreak.domain.HostGroup;
+import com.sequenceiq.cloudbreak.domain.HostMetadata;
+import com.sequenceiq.cloudbreak.domain.InstanceGroup;
+import com.sequenceiq.cloudbreak.domain.InstanceMetaData;
+import com.sequenceiq.cloudbreak.domain.Stack;
+import com.sequenceiq.cloudbreak.domain.Status;
+import com.sequenceiq.cloudbreak.repository.HostGroupRepository;
+import com.sequenceiq.cloudbreak.repository.HostMetadataRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceGroupRepository;
+import com.sequenceiq.cloudbreak.repository.InstanceMetaDataRepository;
+import com.sequenceiq.cloudbreak.repository.ResourceRepository;
+import com.sequenceiq.cloudbreak.repository.StackUpdater;
+import com.sequenceiq.cloudbreak.service.PollingResult;
+import com.sequenceiq.cloudbreak.service.cluster.AmbariClientProvider;
+import com.sequenceiq.cloudbreak.service.cluster.flow.AmbariClusterConnector;
+import com.sequenceiq.cloudbreak.service.events.CloudbreakEventService;
+import com.sequenceiq.cloudbreak.service.stack.StackService;
+import com.sequenceiq.cloudbreak.service.stack.connector.SyncStateFinder;
+
+@Service
+public class StackSyncService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(StackSyncService.class);
+
+    @Inject
+    private StackService stackService;
+    @Inject
+    private StackUpdater stackUpdater;
+    @Inject
+    private CloudbreakEventService eventService;
+    @Inject
+    private InstanceMetaDataRepository instanceMetaDataRepository;
+    @Inject
+    private InstanceGroupRepository instanceGroupRepository;
+    @Inject
+    private ResourceRepository resourceRepository;
+    @Inject
+    private HostGroupRepository hostGroupRepository;
+    @Inject
+    private HostMetadataRepository hostMetadataRepository;
+    @Inject
+    private AmbariClusterConnector ambariClusterConnector;
+    @Inject
+    private AmbariClientProvider ambariClientProvider;
+    @Inject
+    private AmbariHostsRemover ambariHostsRemover;
+    @javax.annotation.Resource
+    private Map<CloudPlatform, SyncStateFinder> syncStateFinders;
+
+    public void sync(Long stackId) throws Exception {
+        Stack stack = stackService.getById(stackId);
+        Set<InstanceMetaData> instanceMetaDatas = instanceMetaDataRepository.findAllWithTerminatedInStack(stackId);
+        String statusReason = "State of the cluster infrastructure has been synchronized.";
+        for (InstanceMetaData instanceMetaData : instanceMetaDatas) {
+            InstanceGroup instanceGroup = instanceMetaData.getInstanceGroup();
+            InstanceSyncState state = syncStateFinders.get(stack.cloudPlatform()).getState(stackId, instanceMetaData.getInstanceId());
+            if (DELETED.equals(state) && !instanceMetaData.isTerminated()) {
+                updateMetadataToTerminatedAndDeregisterFromAmbari(stackId, stack, instanceMetaData, instanceGroup);
```

Comment:
what if this instance was not yet registered in Ambari? e.g.: it was added through the cb-shell to the stack and removed later manually. It can easily be a NPE.
btw it can be 2 separate methods instead of this method name :)